File: C++/level2/라면공장.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <functional>
#include <queue>

using namespace std;

int solution(int stock, vector<int> dates, vector<int> supplies, int k) {
    int sum = stock;
    int answer = 0;
    priority_queue<int> supply;

    reverse(dates.begin(), dates.end());
    reverse(supplies.begin(), supplies.end());

    while(sum < k){
        for(int i=dates.size()-1; i>=0; i--){
            if(dates[i] <= sum){
                supply.push(supplies[i]);
                dates.pop_back();
                supplies.pop_back();
            }
            else
                break;
        }
        sum += supply.top();
        supply.pop();
        answer++;
    }

    return answer;
}

File: JAVA/level2/라면_공장.java
import java.util.*;

class Solution {
    public int solution(int stock, int[] dates, int[] supplies, int k) {
        int answer = 0;
        PriorityQueue<Integer> cache = new PriorityQueue<Integer>(Collections.reverseOrder());
        int day = stock;
        int index = 0;

        while(day < k){
            for(int i=index; i<dates.length; i++){
                if(dates[i] <= day)
                    cache.add(supplies[i]);
                else{
                    index = i;
                    break;
                }

                if(i == dates.length-1)
                    index = i+1;
            }

            int temp = cache.poll();
            day += temp;
            System.out.println(day);
            answer++;
        }

        return answer;
    }
}
